tests/testlib/s2n_ktls_test_utils.c

```diff
@@ -0,0 +1,239 @@
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(conn->recv_io_context);
+    POSIX_ENSURE_REF(msg);
+    POSIX_ENSURE_REF(msg->msg_iov);
+
+    struct s2n_test_ktls_io_stuffer *io_ctx = (struct s2n_test_ktls_io_stuffer *) conn->recv_io_context;
+    POSIX_ENSURE_REF(io_ctx);
+    /* Assert ancillary_buffer is growable, which simplifies IO logic. Blocking IO can
+     * be mocked by restricting the data_buffer. */
+    POSIX_ENSURE(io_ctx->ancillary_buffer.growable, S2N_ERR_IO);
+    io_ctx->invoked_count++;
+
+    /* s2n only receives using msg_iovlen of 1 */
+    MOCK_IO_ENSURE(msg->msg_iovlen == 1);
+
+    uint8_t *buf = msg->msg_iov->iov_base;
+    MOCK_IO_ENSURE_REF(buf);
+
+    /* There is no data available so return blocked */
+    if (!s2n_stuffer_data_available(&io_ctx->ancillary_buffer)) {
+        errno = EAGAIN;
+        return -1;
+    }
+
+    ssize_t total_read = 0;
+    /* updated as partial or multiple records are read */
+    size_t updated_requested_len = msg->msg_iov->iov_len;
+    /* track two record_types since its possible to read multiple records of the same type */
+    *record_type = 0;
+    uint8_t next_record_type = 0;
+    while (*record_type == next_record_type) {
+        /* read record_type and number of bytes available in the next record */
+        MOCK_IO_GUARD(s2n_stuffer_read_uint8(&io_ctx->ancillary_buffer, record_type));
+        uint16_t n_avail = 0;
+        MOCK_IO_GUARD(s2n_stuffer_read_uint16(&io_ctx->ancillary_buffer, &n_avail));
+        POSIX_ENSURE_LTE(n_avail, s2n_stuffer_data_available(&io_ctx->data_buffer));
+
+        size_t n_read = MIN(updated_requested_len, n_avail);
+        MOCK_IO_ENSURE(n_read > 0);
+
+        /* we have already verified that there is more data, so this should succeed */
+        MOCK_IO_GUARD(s2n_stuffer_read_bytes(&io_ctx->data_buffer, buf + total_read, n_read));
+        updated_requested_len -= n_read;
+        POSIX_ENSURE_GTE(updated_requested_len, 0);
+        total_read += n_read;
+
+        /* Handle if we partially read a record */
+        ssize_t remaining_len = n_avail - n_read;
+        if (remaining_len) {
+            POSIX_GUARD_RESULT(s2n_test_ktls_update_prev_header_len(io_ctx, remaining_len));
+        }
+
+        /* if already read the requested amount then break */
+        if (updated_requested_len == 0) {
+            break;
+        }
+        /* Attempt to read multiple records (must be of the same type) */
+        if (updated_requested_len) {
+            int ret = s2n_stuffer_peek_char(&io_ctx->ancillary_buffer, (char *) &next_record_type);
+
+            bool no_more_records = ret < 0;
+            bool next_record_different_type = next_record_type != *record_type;
+
+            if (no_more_records || next_record_different_type) {
+                break;
+            }
+        }
+    }
+
+    return total_read;
+}
+
+S2N_RESULT s2n_test_validate_data(struct s2n_test_ktls_io_stuffer *ktls_io, uint8_t *expected_data, uint16_t len)
```

Comment:
I don't think this needs to be a test function(unless I am missing something). You can just call EXPECT_BYTEARRAY_EQUAL in the test itself.

Reply:
Since we have a stuffer, we need to do a raw_read and ref_check before the compare. Its trivial but adds extra lines. Since we are going to have alot of these and there is already a fn to validate the ancillary data I wanted it to a single liner.